### ISSUE TITLE
Fix `function-calc-no-unspaced-operator` false negatives when combined with properly spaced operator

### ISSUE THREAD

**Issue by ehoogeveen-medweb** [stylelint/stylelint#7181]
### What minimal example or steps are needed to reproduce the bug?

```css
.a {
  /* Error */
  top: calc(5px+3px);
}

.b {
  /* No error */
  top: calc(5px+3px - 2px);
}
```

### What minimal configuration is needed to reproduce the bug?

```json
{
  "rules": {
    "function-calc-no-unspaced-operator": true
  }
}
```

### How did you run Stylelint?

[Demo](https://stylelint.io/demo/#N4Igxg9gJgpiBcIB0BDABMAOgOzWg9AFRoCiATmRGWofjngC4QAO8aYKANmABQCszAB4BqAMxCAlAG4cAXxw4kAIwz0CxAHIQ0MClRp1caJq3ZdeAkeMFoAtGgBMkmdnnYQAGhAAzAJacYDRQAWzhEGEEQ5gCkMABnOM9wCGw-AHMEECwjTBAyAFcAuNy2bLw8XO987DAGXxTbDm5bbAhbarjmFDAYKFsWXRQmMhLjApg1NzckyFTfNIAxKmChzIArOJSk2GZExDK0XLiGAE8Azl9sBlHcziGYY9yPNSPT88uGRpT022OUbCgKDIUBuIDuDAe1xAchAsiAA)

### Which Stylelint-related dependencies are you using?

```json
{
  "stylelint": "^15.10.3",
  "stylelint-config-standard": "^34.0.0"
}
```

### What did you expect to happen?

Rule should report an error for each instance of an unspaced operator.

### What actually happened?

No error is reported if the expression contains at least one properly spaced operator.

This actually caused (well, failed to catch) a real error in our codebase. We had the following code:
```css
.foo {
  top: max(-75px, min(-75px, calc(-0.33 * (100vh - 311px -61px - 50px -64px) + 15px)));
}
```
which is not valid (apparently the `-` in `-61px` and `-64px` is interpreted as the unary minus). It worked for a long time because we were using `cssnano` which changed the spacing and "fixed" the expression - but when I switched to [@csstools/postcss-minify](https://www.npmjs.com/package/@csstools/postcss-minify) which leaves the whitespace alone, the style stopped working.

@romainmenke Thought this might interest you (it's not a bug in @csstools/postcss-minify but rather a bug that was previously hidden by another minifier) :)

**Comment by romainmenke** [comment]
Hi @ehoogeveen-medweb,

It seems that [`postcss-calc`](https://github.com/postcss/postcss-calc) (which is included in `cssnano`) is reducing the calc expression even when the expression is actually invalid?

https://cssnano.co/playground/#eyJpbnB1dCI6Ii5mb28ge1xuICB0b3A6IG1heCgtNzVweCwgbWluKC03NXB4LCBjYWxjKC0wLjMzICogKDEwMHZoIC0gMzExcHggLTYxcHggLSA1MHB4IC02NHB4KSArIDE1cHgpKSk7XG59IiwiY29uZmlnIjoiY3NzbmFuby1wcmVzZXQtZGVmYXVsdCJ9

It might be worth opening an issue there in case this specific case isn't yet reported.

-------

I took a short look into `function-calc-no-unspaced-operator` and it seems that this plugin is also a bit behind on the specification.

It doesn't validate any of the newer math functions.

-------

For this specific issue I would say that the `foundOperatorNode` variable is most likely to blame.

If any operator is found it won't do further checks or warnings.

**Comment by romainmenke** [comment]
I took a brief look and noticed a couple of things :
- the test coverage for this rule is only at ±73%
- there used to be an infinite loop / memory look issue and the changes I was planning for this fix might re-introduce that issue

The current parser that is used also doesn't help much because it just parses everything as a `word` node and it doesn't tokenize, parse correctly (e.g. `35px+35px` is erroneously parsed as a single `word`). But changing it to the `@csstools` tokenizer and parser isn't possible because this rule supports non-standard syntax.

_Changing the underlying parser wouldn't immediately fix this issue, but it would make some of the logic in this plugin simpler to grasp for those who are familiar with the CSS specifications._

------

I think a first step should be to increase the test coverage so that the impact of any possible fix can be reviewed more easily.

**Comment by ybiquitous** [comment]
@romainmenke Nice catch, thank you! I now realize the coverage is not enough. You can work on the `main` branch instead of the `v16` branch if you prefer. 👍🏼

**Comment by jeddy3** [comment]
> But changing it to the `@csstools` tokenizer and parser isn't possible because this rule supports non-standard syntax.

We can prioritise spec compliance over non-standard syntax support for our built-in rules i.e. dropping support for non-standard syntax within a major release is a-okay (either `v16` or another).

I'll label the issue as ready to implement if anyone has time. If preferred, it can be tackled in separate pull requests, e.g. for the test coverage, the change to the tokenzier (and dropping non-standard syntax support) and then fixing the bug.

**Comment by ybiquitous** [comment]
@romainmenke Let me ask how to fix this issue. Do you suggest traversing `calc()` nodes by the tokenizer and parser of `csstools` like this?

https://github.com/stylelint/stylelint/blob/f4c75c911275f7c972f285495c6e037c2731981e/lib/rules/function-no-unknown/index.mjs#L61

**Comment by romainmenke** [comment]
Yes, exactly :)

In theory it could be something like this :
- walk the AST
- set a flag "in calc expression" whenever a function node with one of the math function names is encountered
- for any function or parenthesis block where the flag is set
  - iterate over all child nodes
  - ignore all whitespace and comments
  - warn when sequence is found of "not `+-/*`" -> `number`/`dimension`/`percentage`

If needed the numeric tokens also have a `signCharacter` field that contains `+` or `-` if there was such a character in the original source.

**Comment by ybiquitous** [comment]
@romainmenke Thanks for the detailed answer! I'll try to tackle this problem.

**Comment by Mouvedia** [comment]
> It doesn't validate any of the newer math functions.

@romainmenke check

> [fix the false negatives for round and other calc-sum functions by broadening the meaning of calc to calc-sum, and adjusting the rule description accordingly. As it's fixing false negatives, it'd be a minor release, in line with our semantic-versioning policy.](https://github.com/stylelint/stylelint/issues/5695#issuecomment-1079960257)

**Comment by ybiquitous** [comment]
This issue is reproduced with the latest version, so I've labeled "ready to implement" back. I'm now working on refactoring this rule.

By the way, the inactivity of [`postcss-value-parser`](https://github.com/TrySound/postcss-value-parser) may be one of problems.

**Comment by ehoogeveen-medweb** [comment]
Looks like this rule was overhauled quite thoroughly :) Thanks for all your work!
